include/itkVkMultiResolutionPyramidImageFilter.h

```diff
@@ -32,15 +32,47 @@ namespace itk
 {
 
 /** \class VkMultiResolutionPyramidImageFilter
- * \brief TODO
+ * \brief Creates a multi-resolution pyramid with FFT acceleration
  *
- * TODO
+ * VkMultiResolutionPyramidImageFilter re-implements a framework
+ * for creating an image pyramid as laid out in
+ * MultiResolutionPyramidImageFilter. Conditional logic is added
+ * to preemptively select the optimal image smoothing pipeline
+ * that is expected to give the best performance for different
+ * pyramid levels.
+ *
+ * Separable spatial convolution with DiscreteGaussianImageFilter
+ * runs quickly for small kernel sizes but scales poorly with
+ * increasing kernel size. By contrast ITK FFT convolution accelerated
+ * with a VkFFT GPU backend scales slowly with increasing kernel size
+ * but is typically outperformed by spatial convolution filters
+ * for small kernel sizes.
+ *
+ * VkMultiResolutionPyramidImageFilter allows the user to fix the
+ * kernel size threshold at which a performance tradeoff is expected
+ * between spatial and FFT convolution. The exact threshold depends
+ * on user hardware and can be estimated through benchmarking with
+ * scripts in the ITKVkFFTBackend repository.
+ *
+ * Execution defaults to spatial convolution with the Gaussian kernel.
+ * Parameters can be adjusted so that kernel size thresholds are
+ * anisotropic and so that FFT convolution is selected only if
+ * the kernel surpasses the threshold in some M out of N dimensions.
```

Comment:
We are comparing the runtime for Gaussian blurring by separable spatial convolution to Gaussian blurring by FFT convolution. The two relevant formulas are the ones I quoted above from the "Comparison and Discussion" section of that paper.

In the terms you've presented, the paper suggests that 3D separable spatial convolution will run in `O(X * Y * Z * (i + j + k))` time, while FFT convolution will run in something closer to `O(4 * X * Y * Z * log_2(X))` time.

Of course, a complexity comparison will only give us information about how the implementation scales and does not take parallelism into account. We encounter several confounding factors that will force us to gather parameters from empirical data rather than strictly relying on theory:
- ITK pipelining includes a small amount of variable overhead for setting up mini-pipelines, grafting data, etc
- FFTs typically require a padding stage that scales with the kernel size
- ITK separable spatial Gaussian convolution is multithreaded by default
- FFT convolution runs much faster and scales much slower with increasing image size when accelerated via VkFFT backends as opposed to default Vnl implementation

Does this help at all?

Reply:
Thanks @Leengit , this sounds great. Let me try benchmarking with respect to the ratios you've presented here in the `GaussianBlurBenchmark` notebook and share those empirical findings. Then we can reconsider behavior in this filter and apply necessary code changes.

Reply:
@Leengit Here are the results from plotting `GaussianBlurBenchmark` timing results against corresponding kernel/image ratios. `GaussianBlurBenchmark` tests FFT timings with respect to 1) varied image size and then 2) varied kernel size.

---

### Metric 1: ratio = `(I+J+K) / (log_2(X) + log_2(Y) + log_2(Z))`

Results below are from varying the size of the image and requested output region with a fixed kernel of radius=17 (aka kernel size = [35,35,35]). Under the metric defined above, with a fixed kernel size the ratio value will decrease as image size (and thus runtime) increases. 

In this case I'm not sure that the image size ends up being particularly useful for anticipating smoothing performance comparison. The chart instead illustrates one performance nuance which is that FFT convolution runs faster for this kernel size on nearly all input images, except in the case where the requested region is a subregion centered in the original image such that no padding is required for separable convolution to run.

![image-regions-metric1](https://user-images.githubusercontent.com/40648863/165559951-6d996857-4bca-4fda-9b22-9cffd5dde17e.png)

Results below are from varying the size of the kernel radius with an input image of fixed size 400x400x400 with the largest possible output region requested. The point at which FFT performance intersects spatial performance is close, though not exactly the same, for convolution with a 1D vs a 3D kernel. 1D performance inflection occurs at a metric ratio of approximately 0.6, while 3D performance inflection occurs at approximately 1.0.

![kernel-sizes-metric1](https://user-images.githubusercontent.com/40648863/165562124-2cb3ed99-54e9-46ff-8e8c-a29d66d087d6.png)

---

### Metric 2: ratios `I/log_2(X)`, `J/log_2(Y)`, and `K/log_2(Z)`

Results below are from the first experiment of varied image sizes and requested regions plotted against the second proposed metric, which is a set of one ratio per dimension. Plotting results together shows that there is essentially no difference in shape between each component from this metric and the plots above for Metric 1.

![image-regions-metric2](https://user-images.githubusercontent.com/40648863/165563290-2d4bc068-4432-41d1-9e24-42b685a35899.png)

Results below are from the second experiment of varied kernel sizes plotted against the second proposed metric. Understandably, the 1D kernel shows significantly different results in dimensions 1 and 2 because those kernel dimensions are fixed at size 1 and only dimension 0 is varying. In dimension 0 varying the size of 1D kernel results in a performance tradeoff point at approximately x=2 while varying the size of the 3D kernel the tradeoff point is at approximately x=1.

![kernel-sizes-metric2](https://user-images.githubusercontent.com/40648863/165563821-c6212ab6-c3d7-4e73-a980-d68167e27dd1.png)

---

### Discussion

My takeaways from experiments are as follows:
- Taking `log_2` of each image size results in FFT vs spatial performance with regards to image size changing slowly enough as to possibly be negligible for the decisionmaking case of `VkMultiResolutionPyramidImageFilter`, at least in the case of 3D image smoothing.
- Kernel size has direct, observable impact on the performance comparison between FFT and spatial smoothing.
- The three-component metric (Metric 2) is not demonstrated to necessarily give different or more actionable insight than the one-component metric (Metric 1).
- The one-component metric (Metric 1) gets closer to distilling the decision point into a single threshold to check. In this case the performance inflection point at which FFT convolution outperforms spatial convolution occurs at nearly, though not exactly, the same metric value for the 1D and 3D kernel cases.

### Proposal

I am interested in how we could tailor Metric 1 so that it is 1) more accessible to the user and 2) more closely matches experimental results. Based on the experiments here it might be reasonable to rely less on image size and rely solely on the sum of the kernel lengths `I + J + K` as the scalar threshold value to use in deciding which smoothing procedure to use. I will compare the results of this metric along with those of Metric 1 to see if one better matches experimental performance.

Reply:
Thanks!  Excellent work.

I think that you are saying that for an image of size `400 × 400 × 400`, the two convolution approaches have approximately equal runtime if the kernel size is `8 × 8 × 8`.  Likewise if the kernel size is `14 × 1 × 1`.  Is that right?  I would be interested in the corresponding kernel sizes for a smaller image (maybe `100 × 100 × 100`) and a larger image (maybe `1600 × 1600 × 1600` if that is not too unwieldy). -- but only if that doesn't require too much of your time; I don't want to turn this into more of a research project than it already has become.  (If that is a substantial amount of work, I think we could guesstimate that a threshold near 1 is reasonable enough.)

Does either the FFT or spatial convolution approach treat the 1D kernel specially? -- does either recognize that it doesn't have to convolve in two of the three dimensions and skips code accordingly?

For the fixed-image-size plots, I am little puzzled that the x coordinates of the dots of a color are not evenly spaced.

Reply:
> I think that you are saying that for an image of size 400 × 400 × 400, the two convolution approaches have approximately equal runtime if the kernel size is 8 × 8 × 8. Likewise if the kernel size is 14 × 1 × 1. Is that right? I would be interested in the corresponding kernel sizes for a smaller image (maybe 100 × 100 × 100) and a larger image

Yes, essentially. I've revised my approach and started a new notebook to make it easier to vary both image and kernel size, please take a pass at the plots below when you have time.

> Does either the FFT or spatial convolution approach treat the 1D kernel specially? -- does either recognize that it doesn't have to convolve in two of the three dimensions and skips code accordingly?

This would be good to know, I would have to dig into VkFFT to find the answer there. At a glance ITK's `DiscreteGaussianImageFilter` seems to set up N `NeighborhoodOperatorImageFilter` stages for convolution with an ND kernel without checking whether any kernel size is 1. This could be optimized.

> For the fixed-image-size plots, I am little puzzled that the x coordinates of the dots of a color are not evenly spaced.

Both filters use `itk::GaussianOperator` internally which does not have an option for generating to a specific kernel width but instead takes in a variance, max width, and max error, and samples a Gaussian function accordingly. As such I am limited by the filter interface to passing in different variances and separately computing and reporting what kernel size is being produced.

> I don't want to turn this into more of a research project than it already has become

This is more effort than I was expecting to spend on the metric, but I also understand the hope of making the metric as generalizable as possible. I'm contributing benchmarking examples back to ITKVkFFTBackend as we go in order to demonstrate speedup and encourage user adoption.

---

## Procedure

I generated two parameter lists:
- variance: [  1.77827941   3.32829814   6.22937455  11.65914401  21.82171548
  40.84238653  76.44222743 143.07229892 267.77977837 501.18723363]
  - which maps to kernel radii [2, 3, 4, 6, 8, 10, 14, 20, 27, 32]
- image side length: [ 10  34  58  82 107 131 155 180 204 228 252 277 301 325 350]
  - Note that these lengths were chosen to give a reasonable spread while also being able to re-run the experiment in a reasonable amount of time. In the past I've found that the limit for my GPU is an input image of around 450x450x450 for FFT blurring.

I did the following for each entry in the Cartesian product of the two lists:
- Compute `metric1 = sum(kernel_radius * 3) / sum(log_2(image_length) * 3)
- Compute `metric3 = sum(kernel_radius * 3)
- Construct a random image of size image_length ^ 3
- Time spatial smoothing with the given input image and kernel variance
- Time FFT smoothing with the given input image and kernel variance

## Results

Each plot below shows the ratio of spatial smoothing duration to FFT smoothing duration as a percentage. The "best" metric is the one that gives us the best separation between samples where spatial or fft blurring ran faster. In resulting plots this would appear as a vertical bar corresponding to a metric value set by the user. The ideal case is one where a single metric value separates data into two distinct clouds, one where spatial smoothing is fastest (ratio < (10^2 == 100%)) and one where FFT smoothing is fastest (ratio > 10^2).

### Metric 1: Ratio comparing kernel size and log(image size)

![ratio-metric1](https://user-images.githubusercontent.com/40648863/165621337-4ffacb19-d964-4da4-a472-9ce34b56efd0.png)


### Metric 3: Sum of kernel side lengths

![ratio-metric3](https://user-images.githubusercontent.com/40648863/165621359-55dbfe17-bba8-4aeb-b7cb-4ba2c32852bb.png)

## Discussion

- Neither plot gives a distinct separation between spatial and FFT runtimes to easily inform a decision on which method to use for a given input and parameter set.
- Metric 3 (ratio) appears to be slightly better for separating samples benefiting from FFT convolution than Metric 3 (sum of kernel lengths) does. But, neither is great.

### Followup

Out of curiosity I tried adjusting Metric 1 so that the two terms are multiplied instead of divided:
`Metric4 = (I + J + K) * (log_2(X) + log_2(Y) + log_2(Z))`

The results appear to be much more promising for separating samples based on a scalar metric value. Thoughts on this one?

![ratio-metric4](https://user-images.githubusercontent.com/40648863/165624766-998488de-f4b5-4939-9e59-011c0741cdb5.png)

Reply:
Multiplying them makes sense. Bigger images benefit more from GPU processing. And bigger kernels benefit a lot more from FFT filtering. So it would make sense to try this metric: `m4 = I * J * K * log_2(X) * log_2(Y) * log_2(Z)`.

Reply:
@dzenanz That is true. Bigger images can also benefit from CPU multithreading for spatial blurring, though, and I'm not immediately sure how to directly compare VkFFT GPU speedup to CPU multithreading from a theoretical perspective.

The plot for the metric you proposed, which I am going to call "Metric 5" to keep up the naming scheme, is below. I used a log scale on the x-axis as well this time to make it easier to see trends. It seems about equally difficult to divide the "Metric 4" or "Metric 5" plots into spatial/FFT clusters via a single metric scalar.

![ratio-metric5](https://user-images.githubusercontent.com/40648863/165650399-58c81655-b6c6-489d-9c84-3e727c99bd58.png)

Out of curiosity I tried defining and plotting an additional metric with the raw image lengths rather than logs:

`Metric6 = I * J * K * X * Y * Z`

Surprisingly this gives the strongest relationship so far. A scalar value of around `Metric6 = 10^11` does a good job of dividing samples into two groups where relative speed is < 100% for one group and > 100% for the other group. The plot below is on a log scale, so `log(Metric6)` would probably be a better metric in actuality. One caveat here is that my test script is only using cube images and kernels at the moment, so we are really evaluating `(I ^ 3) * (X ^ 3)`; I will try to expand this to include other sizes.

Any thoughts on why using the actual image size results in a strong correlation while we expected the log of the actual image size to do so?

![ratio-metric6](https://user-images.githubusercontent.com/40648863/165650880-db2b5213-4eb1-4707-8400-e5d6732dc5aa.png)

Reply:
At least at the sizes we are trying the FFT approach has a nearly constant runtime.  (That apparently so much time is being used in the set up and very little time is size dependent sounds like a problem to me, but for another day.)

So, Metric 6 make sense for the spatial convolution _if is isn't separable_ because that approach should have a runtime of `IJKXYZ` in that case.  (However, generally Gaussians _are separable_ -- achievable as three passes each with a 1D kernel -- and should have a runtime of `(I+J+K)XYZ`.  That we apparently aren't seeing that sounds like a problem to me ... though perhaps for another day.)

I think that we should go with Metric 6 (unless you think it is worth the time and effort to first explore `(I+J+K)XYZ`).

Because the appropriate metric could change as we explore these "another day" questions, we might consider not exposing the API for changing the threshold at this time.  Instead we will provide a reasonable default.  That allows us to change the metric in the future without deprecating any API, and we can expose the threshold-changing API once we have a stable metric.

Reply:
@tbirdso Metric 7 and expose the API ... that works for me.  (If there is a way to designate the API as private or experimental, perhaps that would make sense; regardless, I am happy to leave that decision to you.)

Apologies for the long detour; thank you for all the data and stick-to-itness!

Reply:
Updated to use "Metric 7" in most recent set of commits.